### PR TITLE
Global setting to allow/disallow users to force stop a vm

### LINKED ISSUE #6629: Add a permission for force stopping a vm
##### ISSUE TYPE

 * Enhancement Request

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
N/A
~~~

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

N/A

##### SUMMARY

We would like the ability to prevent users to use `forced=true` when calling the `stopVirtualMachine` API.

This is wanted to prevent a scenario where a vm is running twice due to an issued forced stop by a user. This occurs when a force stop is sent to a vm, but the agent cannot handle the stop command (disabled/alert/other). The force stop will just administratively set the vm down, while it is still running. when starting the vm again CS will start on another host since the original is in disabled/alert state.

##### STEPS TO REPRODUCE

- Start a vm on a host.
- Disable cloudstack-agent on the host
- Force stop the vm
- Start the vm
 
~~~

~~~

### PR DESCRIPTION
**Pull request by abh1sar**
### Description

Fixes #6629 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):

Screenshot from 2024-08-22 15-27-42
Screenshot from 2024-08-22 15-27-57
Screenshot from 2024-08-22 15-33-36

### CODE REVIEW

**Review by vishesh92 — APPROVED** [review]
clgtm. didn't test.

**Comment by abh1sar** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/response/CapabilitiesResponse.java [hunk 1]
```diff
@@ -92,6 +92,10 @@ public class CapabilitiesResponse extends BaseResponse {
     @Param(description = "true if users can see all accounts within the same domain, false otherwise")
     private boolean allowUserViewAllDomainAccounts;
 
+    @SerializedName("allowuserforcestopvm")
```
Comment (shwstppr) [inline api/src/main/java/org/apache/cloudstack/api/response/CapabilitiesResponse.java:95]: Better to define it in ApiConstants

File: api/src/main/java/org/apache/cloudstack/api/command/user/config/ListCapabilitiesCmd.java [hunk 2]
```diff
@@ -55,6 +55,7 @@ public void execute() {
         response.setAllowUserExpungeRecoverVM((Boolean)capabilities.get("allowUserExpungeRecoverVM"));
         response.setAllowUserExpungeRecoverVolume((Boolean)capabilities.get("allowUserExpungeRecoverVolume"));
         response.setAllowUserViewAllDomainAccounts((Boolean)capabilities.get("allowUserViewAllDomainAccounts"));
+        response.setAllowUserForceStopVM((Boolean)capabilities.get("allowUserForceStopVM"));
```
Comment (shwstppr) [inline api/src/main/java/org/apache/cloudstack/api/command/user/config/ListCapabilitiesCmd.java:58]: Use key from ApiConstants?

File: server/src/main/java/com/cloud/server/ManagementServerImpl.java [hunk 3]
```diff
@@ -4486,6 +4487,7 @@ public Map<String, Object> listCapabilities(final ListCapabilitiesCmd cmd) {
         capabilities.put("allowUserExpungeRecoverVM", allowUserExpungeRecoverVM);
         capabilities.put("allowUserExpungeRecoverVolume", allowUserExpungeRecoverVolume);
         capabilities.put("allowUserViewAllDomainAccounts", allowUserViewAllDomainAccounts);
+        capabilities.put("allowUserForceStopVM", allowUserForceStopVM);
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/server/ManagementServerImpl.java:4490]: same as previous comment

**Review by shwstppr — APPROVED** [review]
Functionality looks okay, left some comments

File: api/src/main/java/org/apache/cloudstack/api/response/CapabilitiesResponse.java (hunk 1 above)
Comment (abh1sar) [reply api/src/main/java/org/apache/cloudstack/api/response/CapabilitiesResponse.java]: Done

File: api/src/main/java/org/apache/cloudstack/api/command/user/config/ListCapabilitiesCmd.java (hunk 2 above)
Comment (abh1sar) [reply api/src/main/java/org/apache/cloudstack/api/command/user/config/ListCapabilitiesCmd.java]: Done

File: server/src/main/java/com/cloud/server/ManagementServerImpl.java (hunk 3 above)
Comment (abh1sar) [reply server/src/main/java/com/cloud/server/ManagementServerImpl.java]: Done

**Comment by abh1sar** [comment]
@blueorangutan package

**Comment by abh1sar** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 4]
```diff
@@ -5341,6 +5341,13 @@ protected void updateVncPasswordIfItHasChanged(String originalVncPassword, Strin
     public void finalizeExpunge(VirtualMachine vm) {
     }
 
+    private void checkForceStopVmPermission(Account callingAccount) {
+        if (!AllowUserForceStopVm.valueIn(callingAccount.getId())) {
+            logger.error(String.format("Parameter [%s] can only be passed by Admin accounts or when the allow.user.force.stop.vm key is true.", ApiConstants.FORCED));
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:5346]:
```suggestion
            logger.error(String.format("Parameter [%s] can only be passed by Admin accounts or when the allow.user.force.stop.vm config is true for the account.", ApiConstants.FORCED));
```

**Review by sureshanaparti — APPROVED** [review]
minor comment, changes lgtm

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 5]
```diff
@@ -5341,6 +5341,13 @@ protected void updateVncPasswordIfItHasChanged(String originalVncPassword, Strin
     public void finalizeExpunge(VirtualMachine vm) {
     }
 
+    private void checkForceStopVmPermission(Account callingAccount) {
+        if (!AllowUserForceStopVm.valueIn(callingAccount.getId())) {
+            logger.error(String.format("Parameter [%s] can only be passed by Admin accounts or when the allow.user.force.stop.vm config is true for the account.", ApiConstants.FORCED));
```
Comment (bernardodemarco) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:5346]: Just a minor suggestion regarding the `log4j2` syntax.

```suggestion
            logger.error("Parameter [{}] can only be passed by Admin accounts or when the allow.user.force.stop.vm config is true for the account.", ApiConstants.FORCED);
```

**Review by bernardodemarco — APPROVED** [review]
LGTM, manually tested in a local environment, here are the test descriptions.

## Test Descriptions

I created two User accounts, `u1` and `u2`. Then, I set the `allow.user.force.stop.vm` setting to `true` for the `u1` account and to `false` for the `u2` account.

### Tests with `u1` account

#### UI

I selected the option to stop a running VM and, as expected, the `Force` input field was rendered on the confirmation form.

image

#### CloudMonkey

I executed the `stopVirtualMachine` API, with the parameter `forced` set to `true`, and the virtual machine was stopped successfully.

### Tests with `u2` account

#### UI

I selected the option to stop a running VM and, as expected, the `Force` input field was not rendered on the confirmation form.

image

#### CloudMonkey

I executed the `stopVirtualMachine` API, with the parameter `forced` set to `true`, and received the following response:

```bash
(localcloud) 🐱 > stop virtualmachine id="de80cbfc-e736-413d-be78-113e093955de" forced="true"
{
  "account": "u2",
  "accountid": "80cc6bd6-6bc5-43d7-85b9-84557ef7501f",
  "cmd": "org.apache.cloudstack.api.command.user.vm.StopVMCmd",
  "completed": "2024-09-01T13:55:56+0000",
  "created": "2024-09-01T13:55:56+0000",
  "domainid": "6b5335fb-4b6f-11ef-87b4-cec422422af1",
  "domainpath": "ROOT",
  "jobid": "f47b1388-6ed5-4c6b-b6f8-9754fbae3863",
  "jobinstanceid": "de80cbfc-e736-413d-be78-113e093955de",
  "jobinstancetype": "VirtualMachine",
  "jobprocstatus": 0,
  "jobresult": {
    "errorcode": 530,
    "errortext": "Account does not have the permission to force stop the vm."
  },
  "jobresultcode": 530,
  "jobresulttype": "object",
  "jobstatus": 2,
  "userid": "852c7737-b9eb-479c-8613-e366e165b6e2"
}
🙈 Error: async API failed for job f47b1388-6ed5-4c6b-b6f8-9754fbae3863
```

As can be seen, the operation was not completed successfully, which is the expected behavior.

**Pull request merged**